make{,Binary}Wrapper: some cleanups

ncfavier opened the pull request:
Nothing crazy, see individual commits.

Fixes https://github.com/NixOS/nixpkgs/issues/172249, addresses https://github.com/NixOS/nixpkgs/pull/172320#discussion_r869185973

ncfavier committed:
makeBinaryWrapper: move into its own folder

The derivation is complex enough to warrant moving out of
all-packages.nix

ncfavier committed:
makeWrapper,makeBinaryWrapper: introduce explicitly named functions

Because both versions might end up in a derivation's build inputs, it
might be useful to be able to explicitly select which function to use.

ncfavier committed:
wrapGAppsHook: rename argument to makeWrapper

...and pass it makeBinaryWrapper in all-packages.nix

ncfavier committed:
makeBinaryWrapper: remove cc from deps

Fixes https://github.com/NixOS/nixpkgs/issues/172249

ncfavier committed:
makeWrapper: implement --inherit-argv0

For symmetry/interoperability with makeBinaryWrapper. Implemented as
--argv0 '$0'

dotlambda commented on pkgs/tools/security/cryptomator/default.nix:
```diff
@@ -102,7 +102,7 @@ in stdenv.mkDerivation rec {
     maven
     makeWrapper
-    (wrapGAppsHook.override { makeBinaryWrapper = makeWrapper; })
+    (wrapGAppsHook.override { inherit makeWrapper; })
```
```suggestion
    wrapGAppsHookp
```
Just replace makeWrapper with makeShellWrapper.

ncfavier replied:
Indeed, that seems better.

thiagokokada commented on pkgs/test/default.nix:
```diff
@@ -72,6 +72,12 @@ with pkgs;
   dhall = callPackage ./dhall { };
 
   makeWrapper = callPackage ./make-wrapper {};
```
Nit:
```suggestion
  makeWrapper = callPackage ./make-wrapper { };
```

ncfavier replied:
I don't know, I prefer `{}`, the file is mostly full of `{}` and I don't think we have an explicit convention for this anywhere :/

thiagokokada replied:
```
~/Projects/nixpkgs master ⇡
❯ rg -F '{ }' pkgs/top-level/all-packages.nix | wc -l
10181

~/Projects/nixpkgs master ⇡
❯ rg -F '{}' pkgs/top-level/all-packages.nix | wc -l
1101
```

Also this is how it is preferred by both `nixpkgs-fmt` and `nixfmt` formatters.

ncfavier replied:
Fine

thiagokokada commented on pkgs/top-level/all-packages.nix:
```diff
@@ -783,30 +783,7 @@ with pkgs;
-      cc = stdenv.cc;
-      sanitizers = [ ];
-    };
+  makeBinaryWrapper = callPackage ../build-support/setup-hooks/make-binary-wrapper {};
```
Another nit.

```suggestion
  makeBinaryWrapper = callPackage ../build-support/setup-hooks/make-binary-wrapper { };
```

thiagokokada commented on pkgs/build-support/setup-hooks/make-binary-wrapper/make-binary-wrapper.sh:
```diff
@@ -15,17 +15,17 @@ assertExecutable() {
 # --unset       VAR     : remove VAR from the environment
 # --chdir       DIR     : change working directory (use instead of --run "cd DIR")
 # --add-flags   FLAGS   : add FLAGS to invocation of executable
+# TODO(@ncfavier): --append-flags
```
I think this will appear on `make-binary-wrapper` help. It is better to move this somewhere else.

ncfavier replied:
Where are you thinking this will appear? Are those comments parsed by anything?

ncfavier replied:
I'll remove the TODO soon™ anyway.

thiagokokada replied:
> Where are you thinking this will appear? Are those comments parsed by anything?

I am almost sure it is parsed by the help itself. Let me see if I remember how to invoke it.

ncfavier replied:
What help?

thiagokokada replied:
I may be misremembering, but I am almost sure that you can call `wrapProgram` manually inside a shell and it would show this text as a help if you don't pass any arguments.

This is what I am trying to remember how to do.

thiagokokada reviewed (commented):
Some general comments, but looks like a very good clean-up 👏 .

ncfavier commented on pkgs/applications/networking/instant-messengers/discord/linux.nix:
```diff
@@ -24,7 +24,8 @@ stdenv.mkDerivation rec {
     mesa
     nss
-    (wrapGAppsHook.override { makeBinaryWrapper = makeWrapper; })
+    wrapGAppsHook
+    makeWrapper
```
No: https://github.com/NixOS/nixpkgs/blob/519c054c1eb32d5fb68de5375364809a00b3a623/pkgs/applications/networking/instant-messengers/discord/linux.nix#L31

Only one wrapper, the `wrapProgramShell` one (with the arguments from the GApps wrapper).

ncfavier commented:
Failures are due to timeout building LLVM, but based on my tests in https://github.com/NixOS/nixpkgs/pull/172257 this should work on every platform.

thiagokokada reviewed (approved):
Code-wise, LGTM.

Not tested, would like a second opinion here.

bergkvist reviewed (approved):
The changes to makeShellWrapper and makeBinaryWrapper look good to me, assuming you were able to run the tests successfully for x86-linux, x86-darwin and aarch64-linux locally.

ncfavier commented:
I built and ran firefox-bin succesfully. Building discord now, but it takes a while.

ncfavier commented:
Built and ran firefox-bin, discord, vscode, signal, neovim and mpv.

ncfavier committed:
makeBinaryWrapper: add extractCmd

A small shell script that can be used to extract a binary wrapper's
makeCWrapper call from its embedded docstring, without depending on
makeBinaryWrapper.

ncfavier commented:
Firefox still working.

ncfavier committed:
makeShellWrapper: add explicitly named attribute

So that things that use the makeShellWrapper/wrapProgramShell
functions can depend on makeShellWrapper explicitly, which should ease
migration in the future.

ncfavier commented:
I keep thinking of more things 😅 this is just a naming change though.

thiagokokada commented:
Well, tell me when you're ready. I was going to avoid merging this until 22.05 is released, but we still have some time and the changes seem safe-ish.

ncfavier commented:
I think this is fine to merge now.

bergkvist commented:
With the extractCommand stuff you added, the path should probably be escaped properly here (in case it contains spaces):
https://github.com/NixOS/nixpkgs/blob/71637916319afa4aa752fa158f4eca392c937069/pkgs/build-support/setup-hooks/make-binary-wrapper/make-binary-wrapper.sh#L336-L337

Like what is happening here:
https://github.com/NixOS/nixpkgs/blob/71637916319afa4aa752fa158f4eca392c937069/pkgs/build-support/setup-hooks/make-binary-wrapper/make-binary-wrapper.sh#L391

ncfavier commented:
Good catch!

SuperSandro2000 commented on pkgs/test/default.nix:
```diff
@@ -71,7 +71,13 @@ with pkgs;
   dhall = callPackage ./dhall { };
 
-  makeWrapper = callPackage ./make-wrapper {};
+  makeWrapper = callPackage ./make-wrapper { };
+  makeBinaryWrapper = callPackage ./make-binary-wrapper {
+    makeBinaryWrapper = pkgs.makeBinaryWrapper.override {
+      sanitizers = pkgs.lib.optionals (! (pkgs.stdenv.isDarwin && pkgs.stdenv.isAarch64))
+        [ "undefined" "address" ];
+    };
+  };
```
Can we add a short comment why we want this?

ncfavier replied:
Do you mean why we want sanitizers, or why we want them for tests only?

SuperSandro2000 replied:
Why they are tests only and why they are needed to be ovewritten here. That would make the code a lot more understandable.

SuperSandro2000 replied:
You already added a comment. Great.

ncfavier replied:
I added a comment

SuperSandro2000 commented on pkgs/build-support/setup-hooks/make-binary-wrapper/default.nix:
```diff
@@ -0,0 +1,29 @@
+    ++ lib.optional (stdenv.isDarwin && stdenv.isAarch64) darwin.cctools;
+
+  substitutions = {
+    cc = let
+      san = lib.escapeShellArgs (map (s: "-fsanitize=${s}") sanitizers);
+    in "${cc}/bin/cc ${san}";
```
```suggestion
    cc = "${cc}/bin/cc ${lib.escapeShellArgs (map (s: "-fsanitize=${s}") sanitizers)}";
```
Why do we even need other sanitize flags for aarch64?

ncfavier committed:
makeBinaryWrapper: add comment

ncfavier commented:
The `--append-flags` implementation is ready, but I'll make another PR for that.
